File: authipt.c
/*
 * Copyright (C) 1998 - 2007 Bob Beck (beck@openbsd.org).
 * 			2010 Andreas Bertheussen (andreas@elektronisk.org) 
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <err.h>
#include <errno.h>

#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <string.h>

#include <arpa/inet.h>

#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/prctl.h>
#include <fcntl.h>
#include <pwd.h>

#include <sys/param.h>

#include <signal.h>

#include "pathnames.h"

size_t cplen = 0;
char pidfile[MAXPATHLEN];

FILE	*pidfp;
int	pidfd = -1;

char ipsrc[128]; /* xxx.xxx.xxx.xxx\0 */ 
char luser[32];
char proctitle[128];

static int	change_filter(int, const char *, const char *);
static int	change_table(int, const char *);
static void	do_death(int);
static void	need_death(int signo); /* signal handler */
volatile sig_atomic_t	want_death;

int main(int argc, char *argv[]) {
	int lockcnt = 0,n = 0;
	char *cp, *shell;
	struct in6_addr	 ina;
	uid_t uid;
	gid_t gid;
	struct passwd *pw;
	openlog("authipd",LOG_PID|LOG_NDELAY,LOG_DAEMON);
	
	if ((cp = getenv("SSH_TTY")) == NULL) {
		syslog(LOG_ERR, "non-interactive session connection for authpf");
		exit(1);
	}
	
	if ((cp = getenv("SSH_CLIENT")) == NULL) {
		syslog(LOG_ERR, "cannot determine connection source");
		exit(1);
	}
	strncpy(ipsrc, cp, sizeof(ipsrc)); /* fit the stirng into ipsrc */
	if (strlen(ipsrc) < strlen(cp)){
		syslog(LOG_ERR, "SSH_CLIENT variable too long");
		exit(1);
	}

	cp = strchr(ipsrc, ' '); /* Look for the space delimiter after IP address */
	if (cp == NULL) {
		syslog(LOG_ERR, "corrupt SSH_CLIENT variable %s", ipsrc);
		exit(1);
	}
	*cp = '\0';
	
	if (inet_pton(AF_INET, ipsrc, &ina) != 1 &&
		inet_pton(AF_INET6, ipsrc, &ina) != 1) {
		syslog(LOG_ERR, "cannot determine IP from SSH_CLIENT %s", ipsrc);
		exit(1);
	}
	uid = getuid();
	pw = getpwuid(uid);
	if (pw == NULL) {
		syslog(LOG_ERR, "cannot find user for uid %u", uid);
		exit(1);	
	}
	shell = pw->pw_shell; /* Make sure the users shell is set to authipf (user is allowed to run authipf) */
	if (strcmp(shell, PATH_AUTHIPT_SHELL)) {
		//syslog(LOG_ERR, "wrong shell for user %s, uid %u", pw->pw_name, pw->pw_uid);
		// exit(1);  /* TODO: ENABLE THIS */
	}
	
	strncpy(luser, pw->pw_name, sizeof(luser));
	if (strlen(pw->pw_name) > strlen(luser)) {
		syslog(LOG_ERR, "username too long: %s", pw->pw_name);
		exit(1);
	}
	/* The filename to the file for the users IP, e.g. /var/authipt/192.168.2.44 */
	n = snprintf(pidfile, sizeof(pidfile),"%s/%s",
		PATH_PIDFILE,
		ipsrc
		);
	/* a return value of /size/ (sizeof(pidfile)) or more means output was truncated */
	if (n < 0 || (u_int)n >= sizeof(pidfile)) {
		syslog(LOG_ERR, "path to pidfile too long");
		exit(1);
	}
	
	signal(SIGTERM, need_death);
	signal(SIGINT, need_death);
	signal(SIGALRM, need_death);
	signal(SIGPIPE, need_death);
	signal(SIGHUP, need_death);
	signal(SIGQUIT, need_death);
	signal(SIGTSTP, need_death);

	/*
	 * If someone else is already using this ip, then this person
	 * wants to switch users - so kill the old process and exit
	 * as well.
	 *
	 * Note, we could print a message and tell them to log out, but the
	 * usual case of this is that someone has left themselves logged in,
	 * with the authenticated connection iconized and someone else walks
	 * up to use and automatically logs in before using. If this just
	 * gets rid of the old one silently, the new user never knows they
	 * could have used someone else's old authentication. If we
	 * tell them to log out before switching users it is an invitation
	 * for abuse.
	 */
	
	do {
		int save_errno, otherpid = -1;
		char otherluser[32];
		if ((pidfd = open(pidfile, O_RDWR|O_CREAT, 0664)) == -1 ||
		    (pidfp = fdopen(pidfd, "r+")) == NULL) {
			if (pidfd != -1)
				close(pidfd);
			syslog(LOG_ERR, "cannot open or create %s: %s", pidfile, strerror(errno));
			goto die;
		}
		fchmod(pidfd, S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP);
		if (flock(fileno(pidfp), LOCK_EX|LOCK_NB) == 0)
			break;
		save_errno = errno;
		rewind(pidfp);
		if (fscanf(pidfp, "%d\n%31s\n", &otherpid, otherluser) != 2)
			otherpid = -1;
		syslog(LOG_DEBUG, "tried to lock %s, in use by pid %d: %s",
			pidfile, otherpid, strerror(save_errno));

		if (otherpid > 0) {
			syslog(LOG_INFO, "killing prior auth (pid %d) of %s by user %s",
				otherpid, ipsrc, otherluser);
			if (kill((pid_t) otherpid, SIGTERM) == -1) {
				syslog(LOG_INFO, "could not kill process %d: (%m)", otherpid);
			}
		}
		/*
		 * We try to kill the previous process and acquire the lock
 		 * for 10 seconds, trying once a second. if we can't after
 		 * 10 attempts we log an error and give up.
 		 */
 		if (want_death || ++lockcnt > 10) {
			if (!want_death)
				syslog(LOG_ERR, "cannot kill previous authpf (pid %d)", otherpid);
			fclose(pidfp);
			pidfp = NULL;
			pidfd = -1;
			goto dogdeath;
			
		}
		sleep(1);
		/* re-open, and try again. The previous authpf process
		 * we killed above should unlink the file and release
 		 * it's lock, giving us a chance to get it now
		 */
		fclose(pidfp);
		pidfp = NULL;
		pidfd = -1;
	} while (1);

	/* whack the group list */
	gid = getegid();
	if (setgroups(1, &gid) == -1) {
		syslog(LOG_INFO, "setgroups: %s", strerror(errno));
		do_death(0);
	}

	/* TODO: CHECK IF USER IS BANNED */
	/* TODO: CONFIG FILE */

	/* TODO: remove stale rulesets */

	rewind(pidfp);
	fprintf(pidfp, "%ld\n%s\n", (long)getpid(), luser);
	fflush(pidfp);
	(void) ftruncate(fileno(pidfp),ftello(pidfp));

	//syslog(LOG_ERR, "Adding IP address %s", ipsrc);
	if (change_filter(1,luser,ipsrc) == -1) {
		printf("Unable to modify filters\n");
		do_death(0);
	}
	if (change_table(1, ipsrc) == -1){
		printf("Unable to modify table\n");
		change_filter(0,luser,ipsrc);
		do_death(0);
	}
	
	/* revoke privs */

	/*uid = getuid();
	if (setresuid(uid, uid, uid) == -1) {
		syslog(LOG_INFO, "setresuid: %s", strerror(errno));
		do_death(1);
	}*/


	while (1) {
		struct stat sb;
		char *path_message;
		printf("Hello %s.\n", luser);
		printf("You are authenticated from host %s\n.", ipsrc);
		snprintf(proctitle, sizeof(proctitle), "%s@%s", luser, ipsrc);
		prctl(PR_SET_NAME, proctitle, NULL, NULL, NULL);
		/* TODO: rename the process better than this - prctl has a limit of 15 letters */	

		/* TODO: print custom message from file*/
		while(1) {
			sleep(10);
			if (want_death)
				do_death(1);
		}
	}

	return 0;
dogdeath:
	printf("\n\nSorry, this service is currently unavailable due to ");
	printf("technical difficulties\n");
	printf("Your authentication process (pid %ld) was unable to run\n", (long)getpid());
	sleep(180);
die:
	do_death(0);
	return 0;
}

static int change_filter(int add, const char *luser, const char *ipsrc) {
	if (add) {
		//printf("[+] Adding user-specific rules for %s@%s\n", luser, ipsrc);
	} else {
		//printf("[+] Removing user-specific rules for %s@%s\n", luser, ipsrc);

	}
	return(0);
}

static int change_table(int add, const char *ipsrc) {
	/* make sure the table exists */
	pid_t pid;
	gid_t gid;
	int s;
	char *ipstr = NULL;
	char *pargv[5] = {PATH_IPSET, "-N", "authipt", "iphash", NULL};
	
	if (luser == NULL || !luser[0] || ipsrc ==NULL || !ipsrc[0]) {
		syslog(LOG_ERR, "invalid luser/ipsrc");
		goto error;
	}
	
	switch(pid = fork()) {
		case -1: syslog(LOG_ERR, "fork failed");
			goto error;
		case 0: gid = getgid();
			if (setregid(gid,gid) == -1) {
				syslog(LOG_ERR, "setregid: %s", strerror(errno));
			}
			execvp(PATH_IPSET, pargv);
			syslog(LOG_ERR, "exec of %s failed", PATH_IPSET);
			_exit(1); /* abort the child process */
		default: break;/* this is the parent process, continue */
	}
	
	waitpid(pid, &s, 0);

	pargv[3] = ipsrc;
	if (add) {
		//printf("[+] Adding %s to set of authenticated users\n", ipsrc);
		pargv[1] = "-A";
	} else {
		//printf("[-] Removing %s from set of authenticated users\n", ipsrc);
		pargv[1] = "-D";
	}
	switch(pid = fork()) {
		case -1:
			syslog(LOG_ERR, "fork failed");
			goto error;
		case 0: /* This is the child process - execute command */
			gid = getgid();
			if (setregid(gid, gid) == -1) {
				syslog(LOG_ERR, "setregid: %s", strerror(errno));
			}
			execvp(PATH_IPSET, pargv);
			syslog(LOG_ERR, "exec of %s failed", PATH_IPSET);
			_exit(1);
		default: break; /* this is the parent process, continue */
	}
	waitpid(pid, &s, 0);

	return(0);
error:
	return(-1);

}

static void need_death(int signo) {
	want_death = 1; /* Signal through a variable */
}

static void do_death(int active) {
	int ret = 0;
	if (active) {
		change_filter(0, luser, ipsrc);
		change_table(0, ipsrc);
		/* TODO: kill states */

	}

	if (pidfile[0] && pidfd != -1)
		if (unlink(pidfile) == -1)
			syslog(LOG_ERR, "cannot unlink (%s) (%m)",pidfile);
	exit(0);

}

File: authipt.py
# Copyright (c) 2010, Andreas Bertheussen <andreas@elektronisk.org>
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.

# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

import os, sys, time, pwd, grp, fcntl
import syslog, socket, signal, subprocess
import setproctitle, ConfigParser

# termination handler does not exit directly. wantdeath should be checked
# regularly through the program
def needdeath(arg1, arg2):
	global wantdeath
	wantdeath = True
	return

# updaterules() and updateset() invoke ipset and iptables
def updaterules(add, userip, username):
	uprulesfilename = '/etc/authipt/users/%s/uprules' % username
	downrulesfilename = '/etc/authipt/users/%s/downrules' % username

	# verify that user has either both up/downrules, or neither
	if (os.path.isfile(uprulesfilename) ^ os.path.isfile(downrulesfilename)):
		syslog.syslog(LOG_ERR, 'user %s has a lonely \'uprules\' /or/ \'downrules\' file' % username)
		return False
	# check that the rulefile for the requested action exists
	if add == True:
		rulesfilename = uprulesfilename
	else:
		rulesfilename = downrulesfilename
	
	if not os.path.isfile(rulesfilename):
		return True	# no rulefiles exist, nothing more to do here
	
	try:
		rulesfile = open(rulesfilename, 'r')
		rules = rulesfile.readlines()
	except IOError, (errno, strerror):
		syslog.syslog(LOG_ERR, 'could not open and/or read rulefile %s: %s' % (rulesfilename, strerror))
		return False
	
	# replace the rule macros $username, $userip and $userpid with the appropriate values
	rules = ''.join(rules)
	rules = rules.replace('$username', username)
	rules = rules.replace('$userip', userip)
	rules = rules.replace('$userpid', '%i' % os.getpid())
	
	# run iptables-restore, and feed it the rules
	global conf
	try:	iptrestorepath = conf.get('authipt', 'iptrestore')
	except:	iptrestorepath = '/sbin/iptables-restore'

	if not os.path.isfile(iptrestorepath):
		syslog.syslog(LOG_ERR, 'could not find iptables-restore at %s' % iptrestorepath)
		return False

	iptargs = '%s -n' % iptrestorepath
	iptargs = iptargs.split(' ')
	try:
		ipt = subprocess.Popen(iptargs, shell=False, stdin=subprocess.PIPE)
	except OSError, (errno, strerror):
		syslog.syslog(LOG_ERR, 'could not call iptables-restore: %s' % strerror)
		return False
	
	ipt.communicate(rules)	# write rule to stdin of iptables-restore
	return True

def updateset(add, userip):
	global conf
	try:	ipsetpath = conf.get('authipt', 'ipset')
	except:	ipsetpath = '/usr/sbin/ipset'

	if not os.path.isfile(ipsetpath):
		syslog.syslog(LOG_ERR, 'could not find ipset at %s' % ipsetpath)
		return False

	try:	ipsetname = conf.get('authipt', 'ipsetname')
	except:	ipsetname = 'authipt'
	ipsetargs = '%s -N %s iphash' % (ipsetpath, ipsetname)
	ipsetargs = ipsetargs.split(' ')
	if add == True:
		# make sure set exists before adding ip to it
		try:
			ipset = subprocess.Popen(ipsetargs, shell=False, stderr=subprocess.PIPE)
		except OSError, (errno, strerror):
			syslog.syslog(LOG_ERR, 'could not call ipset for table creation: %s' % strerror)
			return False
		
		ipset.communicate()	# waits until ipset terminates
		ipsetargs[1] = '-A'
	else:
		ipsetargs[1] = '-D'
	
	ipsetargs[3] = userip
	try:
		ipset = subprocess.Popen(ipsetargs, shell=False)
	except OSError, (errno, strerror):
		syslog.syslog(LOG_ERR, 'could not call ipset for IP insertion/removal: %s' % strerror)
		return False

	ipset.communicate()
	return True

def do_death(active):
	global pidfile, pidfilename, userip, username
	pidfile.close()
	if os.path.isfile(pidfile.name):
		os.unlink(pidfile.name)
	if active == 1:
		updaterules(False, userip, username)
		updateset(False, userip)
	sys.exit()

	return

def printfile(filename):
	try:
		f = open(filename, 'r')
		for line in f:
			print line,	# comma supresses superfluous \n char
		f.close()
	except IOError:
		pass	# ignore errors
	return

LOG_INFO = syslog.LOG_INFO
LOG_ERR = syslog.LOG_ERR
syslog.openlog('authipt', syslog.LOG_PID|syslog.LOG_NDELAY, syslog.LOG_DAEMON)

conf = ConfigParser.RawConfigParser()

try:	conf.readfp(open('/etc/authipt/authipt.conf'))
except:	pass

# we must be root (uid 0) to run (through sudo)
if  os.getuid() != 0:
	syslog.syslog(LOG_ERR, 'user with uid %i was denied to run authipt (not root)' % os.getuid())
	print 'You are not root, %s.' % pwd.getpwuid(os.getuid()).pw_name
	sys.exit()

try:
	SSH_TTY = os.environ['SSH_TTY']
	SSH_CONNECTION = os.environ['SSH_CONNECTION']
except KeyError:
	syslog.syslog(LOG_ERR, 'SSH_TTY or SSH_CONNECTION were not provided')
	sys.exit()

try:
	userid = int(os.environ['SUDO_UID'])
	username = os.environ['SUDO_USER']
except KeyError:
	syslog.syslog(LOG_ERR, 'SUDO_UID or SUDO_USER were not provided')
	sys.exit()

userpwd = pwd.getpwuid(userid)
if userpwd.pw_name != username:
	syslog.syslog(LOG_ERR, 'SUDO_UID and SUDO_USER do not refer to same user')
	sys.exit()

try:	shellname = conf.get('authipt', 'shell')
except:	shellname = '/bin/authipt'

if not os.path.isfile(shellname):
	syslog.syslog(LOG_ERR, "current required shell %s does not exist, which means noone can log in with it")
	sys.exit()

# only let users with a correct shell, run authipt
if userpwd.pw_shell != shellname:
	syslog.syslog(LOG_ERR, 'denied user %s access because authipt is not the users shell (%s)' % (username, userpwd.pw_shell))
	sys.exit()

# make sure user is in the authipt group
try:	groupname = conf.get('authipt', 'group')
except:	groupname = 'authipt'

try:
	group = grp.getgrnam(groupname)
except KeyError:
	syslog.syslog(LOG_ERR, 'required group %s does not exist', groupname)
	sys.exit()

if username not in group.gr_mem:
	syslog.syslog(LOG_ERR, 'denied user %s access because user is not in expected group %s' % (username, groupname))
	sys.exit()


if SSH_CONNECTION.count(' ') != 3:
	syslog.syslog(LOG_ERR, 'SSH_CONNECTION was malformed')
	sys.exit()

userip = SSH_CONNECTION.split(' ', 2)[0]	# first element is ip

# do a sanity check of the IP address size. NOTE: must be extended for IPv6
if (len(userip) > 16) or (len(userip) < 7):
	syslog.syslog(LOG_ERR, 'IP address \'%s\' was too long or short', userip)
	sys.exit()

try:
	socket.inet_pton(socket.AF_INET, userip)
except socket.error:
	syslog.syslog(LOG_ERR, 'IP address \'%s\' was invalid' % userip )
	sys.exit()

try:	piddir = conf.get('authipt', 'piddir')
except:	piddir = '/var/authipt'

if not os.path.isdir(piddir):
	syslog.syslog(LOG_ERR, 'PID dir %s did not exist')
	sys.exit()

pidfilename = '%s/%s' % (piddir, userip)

signal.signal(signal.SIGTERM, needdeath)
signal.signal(signal.SIGINT, needdeath)
signal.signal(signal.SIGQUIT, needdeath)
signal.signal(signal.SIGALRM, needdeath)
signal.signal(signal.SIGPIPE, needdeath)
signal.signal(signal.SIGHUP, needdeath)
signal.signal(signal.SIGTSTP, needdeath)

try:	confdir = conf.get('authipt', 'confdir')
except:	confdir = '/etc/authipt'

if not os.path.isdir(confdir):
	syslog.syslog(LOG_ERR, 'conf dir %s did not exist')
	sys.exit()

try:	messagetimeout = conf.getint('authipt', 'messagetimeout')
except:	messagetimeout = 60

tries = 0
retry = True
wantdeath = False
while retry:
	retry = False
	try:
		if os.path.isfile(pidfilename):
			pidfile = open(pidfilename, 'rw+')
		else:
			pidfile = open(pidfilename, 'wr+')
	except IOError, (errno, strerror):
		syslog.syslog(LOG_ERR, 'unable to open pidfile %s' % pidfilename)
		do_death(0)
	
	pidfiledesc = pidfile.fileno()
	try:
		fcntl.flock(pidfiledesc, fcntl.LOCK_EX|fcntl.LOCK_NB)
	except IOError:
		syslog.syslog(LOG_ERR, 'pidfile %s is locked, killing owner' % pidfilename)
		retry = True
		pidfile.seek(0)
		lines = pidfile.readlines()
		if len(lines) != 2:
			syslog.syslog(LOG_ERR, 'pidfile %s was corrupt' % pidfilename)
			pidfile.close()
			retry = True
		userpid = int(lines[0])	# first line is pid, second is username
		os.kill(userpid, signal.SIGTERM)

	tries += 1
	if tries == 5:
		syslog.syslog(LOG_ERR, 'gave up grabbing lock for %s' % pidfilename)
		print 'Authentication is unavailable due to technical difficulties'
		printfile('%s/problem' % confdir)
		time.sleep(messagetimeout)
		do_death(0)
	if wantdeath == True:
		do_death(0)
	time.sleep(0.5)

# we have the pidfile lock, and can register a new authentication

banfilename = '%s/users/%s/banned' % (confdir, username)
if os.path.isfile(banfilename):
	print 'Your account is banned from authentication.'
	syslog.syslog(LOG_INFO, 'user %s was rejected due to existing banfile')
	printfile(banfilename)
	time.sleep(messagetimeout)	# give time to read the message
	do_death(0)

try:
	pidfile.seek(0)
	pidfile.writelines(['%i\n' % os.getpid(), username])
	pidfile.truncate()
	pidfile.flush()
except IOError, (errno, strerror):
	syslog.syslog(LOG_ERR, 'could not write pidfile %s: %s' % (pidfilename, strerror))
	do_death(0)

if not updaterules(True, userip, username):
	updaterules(False, userip, username)
	do_death(0)

if not updateset(True, userip):
	do_death(1)

syslog.syslog(LOG_INFO, 'user %s@%s authenticated.' % (username, userip))
print 'Hello %s - you are authenticated from host %s.' % (username, userip)
printfile('%s/motd' % confdir)
setproctitle.setproctitle('authipt: %s@%s' % (username, userip))

while True:	# sleep until killed.
	if wantdeath == True:
		do_death(1)
	time.sleep(5)
